File: packages/worker-core/src/storage.ts
import fs from 'node:fs/promises';
import path from 'node:path';
import os from 'node:os';
import { randomUUID } from 'node:crypto';
import type { BackendApi } from './api-client.js';

/**
 * Einheitlicher Zugriff auf den gemeinsamen Storage.
 *
 * - volume: alle Container teilen sich dasselbe Docker-Volume unter DATA_DIR.
 * - api:    der Worker laeuft auf einem anderen Rechner und holt bzw. schickt
 *           Dateien ueber die interne HTTP-API des Backends.
 *
 * Handler-Code muss den Modus nicht kennen: er arbeitet immer mit lokalen
 * Pfaden, die er ueber pull() bekommt und mit push() zurueckgibt.
 */
export class Storage {
  private readonly tempRoot: string;

  constructor(
    private readonly dataDir: string,
    private readonly mode: 'volume' | 'api',
    private readonly api: BackendApi,
  ) {
    this.tempRoot = mode === 'volume' ? path.join(dataDir, 'temp') : path.join(os.tmpdir(), 'acf');
  }

  /** Verhindert Pfad-Ausbrueche (../) aus dem Datenverzeichnis. */
  private resolveSafe(relativePath: string): string {
    const normalized = path.posix.normalize(relativePath.replace(/\\/g, '/')).replace(/^\/+/, '');
    if (normalized.startsWith('..') || path.posix.isAbsolute(normalized)) {
      throw new Error(`Ungueltiger Storage-Pfad: ${relativePath}`);
    }
    const absolute = path.resolve(this.dataDir, normalized);
    const root = path.resolve(this.dataDir);
    if (absolute !== root && !absolute.startsWith(root + path.sep)) {
      throw new Error(`Storage-Pfad verlaesst das Datenverzeichnis: ${relativePath}`);
    }
    return absolute;
  }

  /** Legt ein leeres Arbeitsverzeichnis fuer einen Job an. */
  async createTempDir(prefix = 'job'): Promise<string> {
    const dir = path.join(this.tempRoot, `${prefix}-${randomUUID().slice(0, 8)}`);
    await fs.mkdir(dir, { recursive: true });
    return dir;
  }

  async removeTempDir(dir: string): Promise<void> {
    if (!dir.startsWith(this.tempRoot)) return;
    await fs.rm(dir, { recursive: true, force: true }).catch(() => undefined);
  }

  /**
   * Stellt sicher, dass eine Storage-Datei lokal als Datei vorliegt, und gibt
   * den lokalen Pfad zurueck.
   */
  async pull(relativePath: string, targetDir?: string): Promise<string> {
    if (this.mode === 'volume') {
      const absolute = this.resolveSafe(relativePath);
      await fs.access(absolute);
      return absolute;
    }
    const dir = targetDir ?? (await this.createTempDir('pull'));
    await fs.mkdir(dir, { recursive: true });
    const target = path.join(dir, path.basename(relativePath));
    const buffer = await this.api.downloadFile(relativePath);
    await fs.writeFile(target, Buffer.from(buffer));
    return target;
  }

  /** Schreibt eine lokal erzeugte Datei in den gemeinsamen Storage. */
  async push(localPath: string, relativePath: string, contentType?: string): Promise<string> {
    if (this.mode === 'volume') {
      const absolute = this.resolveSafe(relativePath);
      await fs.mkdir(path.dirname(absolute), { recursive: true });
      if (path.resolve(localPath) !== absolute) {
        await fs.copyFile(localPath, absolute);
      }
      return relativePath;
    }
    const data = await fs.readFile(localPath);
    await this.api.uploadFile(relativePath, data, contentType);
    return relativePath;
  }

  async writeText(relativePath: string, content: string, contentType = 'text/plain; charset=utf-8'): Promise<string> {
    if (this.mode === 'volume') {
      const absolute = this.resolveSafe(relativePath);
      await fs.mkdir(path.dirname(absolute), { recursive: true });
      await fs.writeFile(absolute, content, 'utf8');
      return relativePath;
    }
    await this.api.uploadFile(relativePath, Buffer.from(content, 'utf8'), contentType);
    return relativePath;
  }

  async readText(relativePath: string): Promise<string> {
    if (this.mode === 'volume') {
      return fs.readFile(this.resolveSafe(relativePath), 'utf8');
    }
    const buffer = await this.api.downloadFile(relativePath);
    return Buffer.from(buffer).toString('utf8');
  }

  async size(localPath: string): Promise<number> {
    const stat = await fs.stat(localPath);
    return stat.size;
  }

  /**
   * Erzeugt den Storage-Pfad nach der Projektkonvention:
   *   projects/<projectId>/<bucket>/<dateiname>
   */
  projectPath(projectId: string, bucket: StorageBucket, fileName: string): string {
    return `projects/${projectId}/${bucket}/${fileName}`;
  }
}

export type StorageBucket =
  | 'scripts'
  | 'images'
  | 'audio'
  | 'videos'
  | 'subtitles'
  | 'thumbnails'
  | 'final'
  | 'published'
  | 'exports';

export const STORAGE_BUCKETS: StorageBucket[] = [
  'scripts',
  'images',
  'audio',
  'videos',
  'subtitles',
  'thumbnails',
  'final',
  'published',
  'exports',
];

File: workers/_shared/acfworker/storage.py
from __future__ import annotations

import os
import shutil
import tempfile
import uuid
from pathlib import PurePosixPath

from .api import BackendApi

BUCKETS = (
    "scripts",
    "images",
    "audio",
    "videos",
    "subtitles",
    "thumbnails",
    "final",
    "published",
    "exports",
)


class Storage:
    def __init__(self, data_dir: str, mode: str, api: BackendApi) -> None:
        self.data_dir = os.path.abspath(data_dir)
        self.mode = mode
        self.api = api
        self.temp_root = os.path.join(self.data_dir, "temp") if mode == "volume" else os.path.join(tempfile.gettempdir(), "acf")

    def _resolve_safe(self, relative_path: str) -> str:
        normalized = str(PurePosixPath(relative_path.replace("\\", "/"))).lstrip("/")
        if normalized.startswith(".."):
            raise ValueError(f"Ungueltiger Storage-Pfad: {relative_path}")
        absolute = os.path.abspath(os.path.join(self.data_dir, normalized))
        if absolute != self.data_dir and not absolute.startswith(self.data_dir + os.sep):
            raise ValueError(f"Storage-Pfad verlaesst das Datenverzeichnis: {relative_path}")
        return absolute

    def create_temp_dir(self, prefix: str = "job") -> str:
        path = os.path.join(self.temp_root, f"{prefix}-{uuid.uuid4().hex[:8]}")
        os.makedirs(path, exist_ok=True)
        return path

    def remove_temp_dir(self, path: str) -> None:
        if not path.startswith(self.temp_root):
            return
        shutil.rmtree(path, ignore_errors=True)

    def pull(self, relative_path: str, target_dir: str | None = None) -> str:
        if self.mode == "volume":
            absolute = self._resolve_safe(relative_path)
            if not os.path.isfile(absolute):
                raise FileNotFoundError(f"Datei fehlt im Storage: {relative_path}")
            return absolute

        directory = target_dir or self.create_temp_dir("pull")
        os.makedirs(directory, exist_ok=True)
        target = os.path.join(directory, os.path.basename(relative_path))
        return self.api.download_file(relative_path, target)

    def push(self, local_path: str, relative_path: str, content_type: str = "application/octet-stream") -> str:
        if self.mode == "volume":
            absolute = self._resolve_safe(relative_path)
            os.makedirs(os.path.dirname(absolute), exist_ok=True)
            if os.path.abspath(local_path) != absolute:
                shutil.copyfile(local_path, absolute)
            return relative_path

        self.api.upload_file(relative_path, local_path, content_type)
        return relative_path

    def write_text(self, relative_path: str, content: str, content_type: str = "text/plain; charset=utf-8") -> str:
        if self.mode == "volume":
            absolute = self._resolve_safe(relative_path)
            os.makedirs(os.path.dirname(absolute), exist_ok=True)
            with open(absolute, "w", encoding="utf-8") as handle:
                handle.write(content)
            return relative_path

        with tempfile.NamedTemporaryFile("w", encoding="utf-8", delete=False, suffix=".tmp") as handle:
            handle.write(content)
            temp_path = handle.name
        try:
            self.api.upload_file(relative_path, temp_path, content_type)
        finally:
            os.unlink(temp_path)
        return relative_path

    def read_text(self, relative_path: str) -> str:
        local = self.pull(relative_path)
        with open(local, "r", encoding="utf-8") as handle:
            return handle.read()

    @staticmethod
    def project_path(project_id: str, bucket: str, file_name: str) -> str:
        return f"projects/{project_id}/{bucket}/{file_name}"
